hack/check-rec-rule-usage.sh

```diff
@@ -4,19 +4,46 @@ set -e
+    exit 1
+fi
+
 findrules="$(cd "$(dirname "${BASH_SOURCE[0]}")" &>/dev/null && pwd)/find-rules.sh"
 
 "$findrules" >"$TMP/rule-groups"
-# extract recording rules
+# Extract recording rules
 tmp/bin/gojsontoyaml -yamltojson <"$TMP/rule-groups" | jq -s '.. | objects | select(.record) | .record' | tr -d '"' | sort -u >"$TMP/rec-rules"
-# and alerting rules
+
+# Extract alerting rules
 tmp/bin/gojsontoyaml -yamltojson <"$TMP/rule-groups" | jq -s '.. | objects | select(.alert) | .expr' | grep "\w\+{" -o | tr -d "{" | sort -u >"$TMP/used-rules"
-# and telemetry client's metrics
-tmp/bin/gojsontoyaml -yamltojson <manifests/*-config.yaml | jq '.data."metrics.yaml"' | sed "s;\\\\n;\n;g" | grep "^-\s'{__name__=\\\\\"" | sed -e "s/^- '{__name__=\\\\\"//" | sed -e "s/\\\\\"[^}]*}'//" | sort | uniq >>"$TMP/used-rules"
 
-# the first grep outputs all recording rules that are used neither in alerts nor telemetry metrics
+# Extract referenced rules in other rules' expressions
+tmp/bin/gojsontoyaml -yamltojson <"$TMP/rule-groups" | jq -s '.. | objects | select(.expr) | .expr' >"$TMP/rule-exprs"
+
+go run -mod=vendor hack/promqlrule.go "$TMP/rule-exprs" >"$TMP/rule-expr-names"
+grep -Fxf "$TMP/rec-rules" "$TMP/rule-expr-names" >>"$TMP/used-rules"
+
+# Extract Telemetry Client's metrics
+# Put plain text matching rule names into used-rules
+tmp/bin/gojsontoyaml -yamltojson <manifests/*-config.yaml | jq '.data."metrics.yaml"' | sed "s;\\\\n;\n;g" | grep "^-\s'{__name__=\\\\\"" | sed -e "s/^- '{__name__=\\\\\"//" | sed -e "s/\\\\\"[^}]*}'//" | sort | uniq >>"$TMP/used-rules"
+# Put regex pattern matching rule names into used-rules-regex
+tmp/bin/gojsontoyaml -yamltojson <manifests/*-config.yaml | jq '.data."metrics.yaml"' | sed "s;\\\\n;\n;g" | grep "^-\s'{__name__=~\\\\\"" | sed -e "s/^- '{__name__=~\\\\\"//" | sed -e "s/\\\\\"[^}]*}'//" | sort | uniq >"$TMP/used-rules-regex"
+# The first grep outputs all recording rules that are used neither in alerts nor telemetry metrics
 # and then greps for the remaining rules in the dashboard defs and outputs the ones that are not found
-grep -Fxvf "$TMP/used-rules" "$TMP/rec-rules" | while read -r r; do grep "$r" assets/grafana/dashboard-definitions.yaml -q || echo "$r"; done >"$TMP/unused-rules"
+grep -Fxvf "$TMP/used-rules" "$TMP/rec-rules" | while read -r r; do grep "$r" assets/grafana/dashboard-definitions.yaml -q || echo "$r"; done >"$TMP/unused-rules-fixstr"
+grep -Exvf "$TMP/used-rules-regex" "$TMP/unused-rules-fixstr" >"$TMP/unused-rules-cmo"
+
+# Find out the rules used in console. The console TypeScript codes contain reference to rule names by exact text matching.
+find "${CONSOLE_DIR}" -type f -regex ".*\.ts[x]*" | while read -r -d '' filename; do
+    grep -Ff "$TMP/unused-rules-cmo" "$filename" || true
+done | sed 's/[^a-zA-Z0-9_]/\n/g' | sort | uniq | grep -Fxf "$TMP/unused-rules-cmo" || true >"$TMP/used-rules-console"
```

Comment:
that part doesn't seem to work because pod:container_cpu_usage:sum is listed in the unused rules while it's used by the console code.